### ISSUE TITLE
https protocol is not supported

### ISSUE THREAD

**Issue by ediri** [KostyaSha/yet-another-docker-plugin#13]
I enter the URL for the UCP in the Field Docker URL. UCP is always with SSL encryption. And i get following error message:

```
com.github.kostyasha.yad.docker_java.org.apache.http.conn.UnsupportedSchemeException: https protocol is not supported
com.github.kostyasha.yad.docker_java.javax.ws.rs.ProcessingException: com.github.kostyasha.yad.docker_java.org.apache.http.conn.UnsupportedSchemeException: https protocol is not supported
    at com.github.kostyasha.yad.docker_java.com.github.dockerjava.jaxrs.connector.ApacheConnector.apply(ApacheConnector.java:488)
    at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:246)
    at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.client.JerseyInvocation$2.call(JerseyInvocation.java:683)
    at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.internal.Errors.process(Errors.java:315)
    at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.internal.Errors.process(Errors.java:297)
    at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.internal.Errors.process(Errors.java:228)
```

i thougth yet-another-docker-plugin supports https?

**Comment by ediri** [comment]
are there any updates?

**Comment by KostyaSha** [comment]
Sorry, i somehow didn't get notification for this issue.

**Comment by KostyaSha** [comment]
@ediri yes, it should support TLS auth against docker daemon. What is `UCP`?

**Comment by KostyaSha** [comment]
Found https://github.com/docker-java/docker-java/issues/419 this may happen if docker connector picked docker environment variables. I will try check this case.

**Comment by KostyaSha** [comment]
Or because of using `unix://` url...

**Comment by ediri** [comment]
UCP is the Docker Universal Control Plane

**Comment by ediri** [comment]
This is from my config.xml file

`  <clouds>
    <com.github.kostyasha.yad.DockerCloud plugin="yet-another-docker-plugin@0.1.0-alpha7">
      <name>Docker Swarm</name>
      <provisionedImages/>
      <templates class="empty-list"/>
      <containerCap>50</containerCap>
      <connector>
        <serverUrl>https://ucp.ww-intern.de:2376</serverUrl>
        <connectTimeout>10000</connectTimeout>
        <readTimeout>0</readTimeout>
        <credentialsId></credentialsId>
      </connector>
    </com.github.kostyasha.yad.DockerCloud>
  </clouds>
`
When i press "Test connection" i get following error:

`com.github.kostyasha.yad.docker_java.javax.ws.rs.ProcessingException: com.github.kostyasha.yad.docker_java.org.apache.http.conn.UnsupportedSchemeException: https protocol is not supported
	at com.github.kostyasha.yad.docker_java.com.github.dockerjava.jaxrs.connector.ApacheConnector.apply(ApacheConnector.java:488)
	at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:246)
	at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.client.JerseyInvocation$2.call(JerseyInvocation.java:683)
	at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.internal.Errors.process(Errors.java:228)
	at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:424)
	at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:679)
	at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:408)
	at com.github.kostyasha.yad.docker_java.org.glassfish.jersey.client.JerseyInvocation$Builder.get(JerseyInvocation.java:308)
	at com.github.kostyasha.yad.docker_java.com.github.dockerjava.jaxrs.VersionCmdExec.execute(VersionCmdExec.java:25)
	at com.github.kostyasha.yad.docker_java.com.github.dockerjava.jaxrs.VersionCmdExec.execute(VersionCmdExec.java:12)
	at com.github.kostyasha.yad.docker_java.com.github.dockerjava.jaxrs.AbstrSyncDockerCmdExec.exec(AbstrSyncDockerCmdExec.java:23)
	at com.github.kostyasha.yad.docker_java.com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:33)
	at com.github.kostyasha.yad.DockerConnector$DescriptorImpl.doTestConnection(DockerConnector.java:159)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.kohsuke.stapler.Function$InstanceFunction.invoke(Function.java:298)
	at org.kohsuke.stapler.Function.bindAndInvoke(Function.java:161)
	at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:96)
	at org.kohsuke.stapler.MetaClass$1.doDispatch(MetaClass.java:121)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:53)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.MetaClass$6.doDispatch(MetaClass.java:249)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:53)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:649)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:727)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:132)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:123)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:49)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:76)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:81)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:610)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:170)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:98)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:408)
	at org.apache.coyote.ajp.AjpProcessor.process(AjpProcessor.java:193)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:607)
	at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:313)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.github.kostyasha.yad.docker_java.org.apache.http.conn.UnsupportedSchemeException: https protocol is not supported
	at com.github.kostyasha.yad.docker_java.org.apache.http.impl.conn.HttpClientConnectionOperator.connect(HttpClientConnectionOperator.java:99)
	at com.github.kostyasha.yad.docker_java.org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:314)
	at com.github.kostyasha.yad.docker_java.org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:357)
	at com.github.kostyasha.yad.docker_java.org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:218)
	at com.github.kostyasha.yad.docker_java.org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:194)
	at com.github.kostyasha.yad.docker_java.org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:85)
	at com.github.kostyasha.yad.docker_java.org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:108)
	at com.github.kostyasha.yad.docker_java.org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:186)
	at com.github.kostyasha.yad.docker_java.org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:72)
	at com.github.kostyasha.yad.docker_java.com.github.dockerjava.jaxrs.connector.ApacheConnector.apply(ApacheConnector.java:441)
	... 71 more`

**Comment by KostyaSha** [comment]
Did you specified tls credentials?

**Comment by KostyaSha** [comment]
Btw, please update to latest version. It should be available in jenkins update center.

**Comment by ediri** [comment]
I would like that its TLS should be configurable and disabled. As UCP creates a self-signed certificiate

**Comment by KostyaSha** [comment]
You must specify client cert that is your key. AFAIK

**Comment by KostyaSha** [comment]
@ediri and how do you connecting to UCP using standard docker client?

**Comment by wallnerryan** [comment]
i am getting the same

```
com.github.kostyasha.yad.docker_java.org.apache.http.conn.UnsupportedSchemeException: https protocol is not supported (show details)
```

**Comment by KostyaSha** [comment]
that what underlying libraries return, and you need paste the whole content of the keys.

**Comment by wallnerryan** [comment]
@KostyaSha thanks, just figured i'd put it here as it through me for a loop for a little.

**Comment by kyhavlov** [comment]
I ran into this as well, to set up auth with certs I converted mine to PKCS#12 format as shown in this post: http://blog.bananacoding.com/blog/how-to-connect-to-the-docker-remote-api-using-kitematic-and-boot2docker-for-mac, and then added it as a Certificate credential in jenkins.

It would be good to give a different error message in this case; "https is not supported" hides the real error message and almost made me think I wouldn't be able to use the plugin at all for lack of TLS support.

**Comment by chiarag** [comment]
I attempted to use this plugin because I ran into issue 615 with the main Docker-Plugin (the HostConfig deprecated 1.12 vs 1.10).

But I am getting the https unsupported protocol. I have TLS Verify. I have a PKCS cert that was working with the main docker-plugin. The test connection was returning the version successfully in the docker-plugin. In this plugin, I get the error.

We're running Jenkins on a Mac, so I can't use the unix:// route (give me the "only for linux" error). I tried TCP instead but it just hangs, with "Testing". Network Utility port scan indicates it's open.

**Comment by KostyaSha** [comment]
I fixed some routines related to keys... I also remember that tried install UCP and realised that their installation process isn't easy. 
There were also changes in https vs tcp naming scheme, that's why it has tls checkbox. 
Could you specify tcp:// as connection in configuration and try again? About 1.12 issue (that is still not released), set 1.23 in API version field and it should work again.

**Comment by chiarag** [comment]
With this workaround plugin (which I'm running the latest version, build 14), I've tried tcp as you've suggested, prefaced with "tcp://" - it doesn't actually let you leave off the protocol, you get an index out of bounds type error. It still hangs. I've tried leaving with and without TLS Verify, with and without the PKCS cert. Does tcp require a different certificate set up than https?

Adding the API version doesn't really fix it because with the Yet Another Docker Plugin, I can't even get the Test Connection to work (so specifying the version doesn't do anything - I still get unsupported protocol for https). With the Docker Plugin, the Test Connection works but then I get the 1.12 issue when it tries to create a container. The original Docker Plugin has no way to specify the API version so setting 1.23 won't work there.

**Comment by KostyaSha** [comment]
`tcp` is how docker names http/https  based connection (not socket based), in docker-java-3.0 it was changed to follow docker CLI terminology. yad-plugin is based on docker-java-3.0 🏧  so followed connection name change "http" -> "tcp" missed update for help page is in master.

**Comment by KostyaSha** [comment]
How i can reproduce this issue? Could you generate pair of credentials (client and server) for 192.168.99.X connection?

**Comment by chiarag** [comment]
I see what you are saying about the tcp name being synonymous with http in docker. Unfortunately, YAD tcp:// just hangs when set up like the docker-plugin https configuration (so I have the same IP and port, with the same Jenkins Credential Selection).

Our Jenkins is on a Mac running El Capitan (which I think is fairly different than the norm). I am using a docker-machine (so, I create a default machine with the standard settings - didn't customize anything) which automatically generates keys/certs and assigns the port. In order to get the docker-plugin to work (at least the Test Connection part), I had to generate the PKCS file (p12) from the certs/key and set that up within Jenkins Credentials Manager. At that point, I am able to use the REST api from a browser and the docker-plugin is able to get the version back from Docker. 

The PKCS file is located on the Jenkins machine. It's all local, but I'm forced to use TCP/HTTP because it's not linux. :-/

Should they not be installed concurrently? Is the fact that Docker Plugin is installed alongside YAD an issue?

**Comment by KostyaSha** [comment]
yad is fully isolated and using newer docker-java library, also it uses netty instead jetty now. I almost installed UCP, could you guide me how i can create certs?

**Comment by KostyaSha** [comment]
https://docs.docker.com/ucp/install-sandbox/#/step-10-download-a-client-bundle this certs?

**Comment by chiarag** [comment]
I followed this guide on creating a Docker Machine. 

https://docs.docker.com/machine/get-started/

When you do:

`docker-machine create --driver virtualbox default`

It creates the cert, key, etc.

**Comment by chiarag** [comment]
I'm not using UCP. I would think that shouldn't be necessary to get this working?

**Comment by chiarag** [comment]
I just realized this issue is talking about UCP but I don't think that it is related to UCP only. Should I create a different issue as it's not exactly what the original description is?

**Comment by chiarag** [comment]
I was able to use socat on the mac to get something working in a somewhat hokey way with YAD - bypassing the TLS part and the docker-machine.

Still seems like something is not right here between the two - why the Test Connection works on Docker Plugin with https but it does not with YAD.

**Comment by KostyaSha** [comment]
socat is needed for bypassing socket connection, i will verify UCP and toolbox configuration tomorrow (i have integration test that passed for credentials).

**Comment by KostyaSha** [comment]
UCP has multiple certs, maybe that's was an issue. For toolbox-connection i reproduced some fail. Will try fix tomorrow.

**Comment by KostyaSha** [comment]
@chiarag bug is in this "test connection" button! Try save configuration and just run job.

For UCP i found one place where probably docker-java should iterate on multiple certs from single file. WIP.

**Comment by chiarag** [comment]
hahaha, well, that explains it! Thank  you for tracking it down!

I went ahead with the socat solution (restricted to local access) for now, since everything is on the same box and the need for certificates when talking locally seems a little silly.

**Comment by KostyaSha** [comment]
@chiarag "test connection" should be fixed in latest version. 
Multiple certs still investigating.

**Comment by KostyaSha** [comment]
Got some free time. Checking again, is it still actual?

**Comment by chiarag** [comment]
We were able to move onto a linux box so I no longer have to connect via https (and I can use the original docker plugin). So I can't validate this for you on my old set up.

**Comment by KostyaSha** [comment]
@chiarag so you connected via docker.sock?

**Comment by chiarag** [comment]
Yes, docker.sock with the main Docker Plugin.

**Comment by KostyaSha** [comment]
Ok, will be merged soon with #82

**Comment by KostyaSha** [comment]
@chiarag added, released. 

@ediri as i understand multiple certs UCP issues is still not solved?
